Feature/kafkasink - Add a kafka sink

hutcheb opened the pull request.

hutcheb committed:
Wrote first value from Kafka

Still need to:-
- check error handling
- check if there is a more efficent way instead of opening closing 
connection on every write.

hutcheb committed:
Removed request connectionstring, added expires field

- Add an expires field which allows us to discard a write request if it 
is too old.

- removed the connection string from the request, it is already defined 
in the config file.

- Added handling/non handling of errors so that requests are discarded 
if an error occurs.

- Added sample sink config file, need to fix topics field.

hutcheb committed:
Documentation and multiple tasks can be configured.

Added topic check when having multiple tasks so that each task will only 
read from one topic.

ottobackwards commented:
not sure what is up with the conflicts?

Also, do we have tests for this?  kafka has test server classes etc, I can point you to some examples

ottobackwards commented:
I think the tests stuff should be a follow on, please open an issue for adding kafka integration tests

hutcheb committed:
Merge branch 'develop' into feature/kafkasink

hutcheb committed:
Kafka connect 

- Minor changes to source connector allow for empty config. This allows 
to add connectors to Kafka that don't have any source or jobs config.

- Add Confluent maven plugin to package connector for Confluent Hub.

- Changed example configs to use the simulated driver instead of S7. 
This allows for a defaut connector to be added that works without 
external PLCs.

- Removed the ConfigDef define calls, this allows for an empty connector 
to be created. I am also having trouble with the Config.Type.LIST types 
resultig in a NullPointerException when creating connectors in 
Confluent.

hutcheb committed:
Kafa Sink Connector

- Uses an Alias for writing same as the source connector.

- Tested Avro convertors

- Backoff timeouts for sink connector.

Outstanding

- Issues with Simulated parsing from STRING to PlcValue

- Have to close the connection for the sink after every poll.

hutcheb committed:
Started to add the documentation to the website

hutcheb committed:
Started to refactor the source connector validation logic

hutcheb committed:
Refactored Source Validate functions.

- Relying on the Config classes now to parse their own ConfigDef's. As 
well as additional validations that aren't picked up by these.

- Improved support for reporting validation errors

hutcheb committed:
Refactored Sink Connector Validation and added test cases

hutcheb committed:
Testing Simulated Driver

hutcheb committed:
Merge branch 'feature/plc4go' into feature/kafkasink

hutcheb committed:
Test

hutcheb committed:
Merge remote-tracking branch 'origin/feature/plc4go' into feature/kafkasink

hutcheb committed:
Fixed Simulated reading/writing of lists

hutcheb committed:
Kafka Documentation and cleaning up.

- Minor fix for the simulated driver to handle arrays.

hutcheb commented:
Worked through the Confluent checklist as well as added tests and documentation.

- The schema for the source has remained unchanged, I have brought the sink connector schema in line with the source connector.
- I have kept the source connector names for the tags as 'fields'. So we don't break people who are using the previous version.
- Also included are fixes for the simulated driver so it can handle arrays.

chrisdutz commented:
As far as I know nobody is publicly using it ... :-/

So I would say: If it would be better to change things ... do it now before we're actually listed in the Confluent Hub

hutcheb commented:
I've been going back and froth between changing the schema for the source connector and the initial reason for it was so that we don't change the schema. Avro and the schema registry do a great job of handling changes to the schema when we add/remove fields.

Considering another case though, if we have a report by exception device that we only want to send the fields that have changed. We don't want to have to send all the fields we have specified in the config file. But only the values that have changed.

I think changing it so we have an array of [field, value] pairs would be a better option.

chrisdutz commented:
This is definitely a valid scenario, so I would suggest to change it.

hutcheb committed:
Set each Plc field to optional

so we don't have to update each field on every scan

hutcheb commented:
It might be a good idea to move all of the fields into their own property. this way we won't run into issues with having pre-configured names such as timestamp conflicting with field names.

chrisdutz commented:
What do you mean with "their own property"? If it's "their own configuration" and hereby their own kafka connect "instance" ... I strongly object. The usecase where we first used it was to collect 2600 datapoints on each PLC ... having 2600 configurations to manage just will not work.

hutcheb commented:
No sorry, something like:-
{
  "fields": {
                   "running": false,
                   "conveyorStop": true,
                    ......
                },
  "timestamp": 1600002002020
}

So that the user defined fields are separate to the timestamp field. This way the user can add a tag 'timestamp' and not have to worry about conflicting with the timestamp field. It also allows us to add optional fields in the future without having to worry about conflicting with peoples user defined fields.

hutcheb committed:
Updated Fields Struct in Schema.

- All Fields are in thier own fields struct.
- We are now able to producer from the source and directly parse it in a 
sink connector.

hutcheb committed:
Documentation update

hutcheb commented:
Should be in a position to merge. I'll leave it open for comments for a day or so.

chrisdutz commented on plc4j/drivers/simulated/src/main/java/org/apache/plc4x/java/simulated/connection/SimulatedDevice.java:
```diff
@@ -97,12 +97,12 @@ public void set(SimulatedField field, PlcValue value) {
                         }
                 }
-                System.out.printf("TEST PLC RANDOM [%s]: %s%n", field.getName(), value.getString());
+                System.out.printf("TEST PLC RANDOM [%s]: %s%n", field.getName(), value.toString());
```
We probably shouldn't be using System.out.prlint stuff

chrisdutz commented on plc4j/integrations/apache-kafka/README.md:
```diff
@@ -60,7 +66,7 @@ In order to start a Kafka Connect system the following steps have to be performe
         bin/kafka-console-consumer.sh --bootstrap-server localhost:9092 --topic test --from-beginning
 
-### Start a Kafka Connect Worker (Standalone)
+### Start a Kafka Source Connect Worker (Standalone)
```
The name of the product is called "Kafka Connect" ... as this is a Source for Kafka Connect, I'd probably name it "Kafka Connect Source Worker"

chrisdutz commented on plc4j/integrations/apache-kafka/README.md:
```diff
@@ -76,7 +82,7 @@ If you want to debug the connector, be sure to set some environment variables be
 In this case the startup will suspend till an IDE is connected via a remote-debugging session.
 
-### Start Kafka Connect Worker (Distributed Mode)
+### Start Kafka Source Connect Worker (Distributed Mode)
```
Same as above

chrisdutz commented on plc4j/integrations/apache-kafka/README.md:
```diff
@@ -95,3 +101,52 @@ The configuration of the Connectors is then provided via REST interface:
     "tasks.max":"1", "file":"test.sink.txt", "topics":"connect-test" }}' http://localhost:8083/connectors
+
+
+### Start a Kafka Sink Connect Worker (Standalone)
```
Here, the thing should be named "Kafka Connect Sink Worker"

chrisdutz commented on plc4j/integrations/apache-kafka/pom.xml:
```diff
@@ -33,8 +33,17 @@
   </properties>
 
+  <repositories>
+    <repository>
```
As this is used to download the plugin, we need to also specify a "pluginRepository" (Just a duplicate with different element name) ... otherwise the build will fail for people who haven't downloaded the plugin before.

chrisdutz reviewed (changes requested):
I didn't look into the details of the code-changes themselves as I have almost no experience with Kafka and Kafka Connect ... just responded to some build-related and naming issues.

Especially the pluiginRepository issue should be addressed before merging.

hutcheb committed:
Update after Chris's comments

hutcheb committed:
Minor changes

chrisdutz commented on plc4j/integrations/apache-kafka/README.md:
```diff
@@ -1,82 +1,190 @@
-2) Start Zookeeper: 
-        
-        bin/zookeeper-server-start.sh config/zookeeper.properties 
+2) Start Zookeeper:
+
+        bin/zookeeper-server-start.sh config/zookeeper.properties
```
I even heard that new versions of Kafka work without Zookeeper ... we should check this and perhaps update this section later on

chrisdutz reviewed (approved):
All looks good to me (With the actual code changes I 100% have faith in what you did) ... thanks for your hard work

hutcheb committed:
Found issue with the scraper continuing to run when connector deleted.
